packages/plugin-ext/src/plugin/telemetry-ext.ts

```diff
@@ -0,0 +1,373 @@
+        return this._isUsageEnabled;
+    }
+
+    set isUsageEnabled(isUsageEnabled: boolean) {
+        if (this._isUsageEnabled !== isUsageEnabled) {
+            this._isUsageEnabled = isUsageEnabled;
+            this.onDidChangeEnableStatesEmitter.fire(this);
+        }
+    }
+
+    get isErrorsEnabled(): boolean {
+        return this._isErrorsEnabled;
+    }
+
+    set isErrorsEnabled(isErrorsEnabled: boolean) {
+        if (this._isErrorsEnabled !== isErrorsEnabled) {
+            this._isErrorsEnabled = isErrorsEnabled;
+            this.onDidChangeEnableStatesEmitter.fire(this);
+        }
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    logUsage(eventName: string, data?: Record<string, any | TelemetryTrustedValue<any>>): void {
+        if (!this.telemetryEnabled || !this.isUsageEnabled) {
+            return;
+        }
+        this.logEvent(eventName, data);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    logError(eventNameOrException: string | Error, data?: Record<string, any | TelemetryTrustedValue<any>>): void {
+        if (!this.telemetryEnabled || !this.isErrorsEnabled || !this.sender) {
+            // no sender available or error shall not be sent
+            return;
+        }
+        if (typeof eventNameOrException === 'string') {
+            this.logEvent(eventNameOrException, data);
+        } else {
+            this.sender.sendErrorData(eventNameOrException, data);
+        }
+    }
+
+    dispose(): void {
+        if (this.sender?.flush) {
+            let tempSender: TelemetrySender | undefined = this.sender;
+            this.sender = undefined;
+            Promise.resolve(tempSender.flush!()).then(tempSender = undefined);
+        } else {
+            this.sender = undefined;
+        }
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private logEvent(eventName: string, data?: Record<string, any>): void {
+        // No sender means likely disposed of, we should no-op
+        if (!this.sender) {
+            return;
+        }
+        data = mixInCommonPropsAndCleanData(data || {}, this.options?.additionalCommonProperties, this.options?.ignoreBuiltInCommonProperties ? undefined : this.commonProperties);
+        this.sender?.sendEventData(eventName, data);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private getCommonProperties(): Record<string, any> {
+        return [];
+    }
+}
+
+interface TelemetrySender {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    sendEventData(eventName: string, data?: Record<string, any>): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    sendErrorData(error: Error, data?: Record<string, any>): void;
+    flush?(): void | Thenable<void>;
+}
+
+// copied and modified from https://github.com/microsoft/vscode/blob/1.76.0/src/vs/workbench/api/common/extHostTelemetry.ts
```

Comment:
I believe we need an IP check for the copied code.

Reply:
The code was copied from core vscode. It is already duplicated in some other places in the tool. Do we need to get an IP check for every line duplicated from the same base? 
I am not yet used with the current policy in the project. I only checked the results of https://github.com/eclipse-theia/theia/actions/runs/4802461092/jobs/8545944821?pr=12453

Reply:
@rschnekenbu do you have a link to a previous IP Check (CQ) where we copied from that file(s) at that given version of VS Code?

Reply:
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/7970 was created to track down the duplicated code from vscode. I will update here as soon as I have some results from the scan.